marimo-tutorials/artificial-intelligence/deep-learning/activation-functions.py

```diff
@@ -0,0 +1,382 @@
+        formula = ""
+        
+        if func_name == "sigmoid":
+            y = 1 / (1 + np.exp(-x))
+            formula = r"$$\sigma(x) = \frac{1}{1 + e^{-x}}$$"
+        elif func_name == "tanh":
+            y = np.tanh(x)
+            formula = r"$$tanh(x) = \frac{e^x - e^{-x}}{e^x + e^{-x}}$$"
+        elif func_name == "relu":
+            y = np.maximum(0, x)
+            formula = r"$$ReLU(x) = max(0, x)$$"
+        elif func_name == "elu":
+            alpha = 1.0  # Default alpha value for ELU
+            y = np.where(x > 0, x, alpha * (np.exp(x) - 1))
+            formula = r"$$ELU(x) = \begin{cases} x & \text{if } x > 0 \\ \alpha(e^x - 1) & \text{otherwise} \end{cases}$$"
+        elif func_name == "prelu":
+            alpha = 0.01
+            y = np.where(x > 0, x, alpha * x)
+            formula = r"$$PReLU(x) = \begin{cases} x & \text{if } x > 0 \\ \alpha x & \text{otherwise} \end{cases}$$"
+        elif func_name == "leaky_relu":
+            alpha = 0.01
+            y = np.where(x > 0, x, alpha * x)
+            formula = r"$$LeakyReLU(x) = \begin{cases} x & \text{if } x > 0 \\ \alpha x & \text{otherwise} \end{cases}$$"
+        elif func_name == "selu":
+            # SELU parameters
+            alpha = 1.6732632423543772
+            scale = 1.0507009873554804
+            y = scale * np.where(x > 0, x, alpha * (np.exp(x) - 1))
+            formula = r"$$SELU(x) = \lambda \times ELU(x)$$"
+        elif func_name == "softplus":
+            y = np.log1p(np.exp(x))
+            formula = r"$$Softplus(x) = log(1 + e^x)$$"
+        elif func_name == "softsign":
+            y = x / (1 + np.abs(x))
+            formula = r"$$Softsign(x) = \frac{x}{1 + |x|}$$"
+        elif func_name == "hard_sigmoid":
+            y = np.clip(0.2 * x + 0.5, 0, 1)
+            formula = r"$$HardSigmoid(x) = \begin{cases} 0 & \text{if } x \leq -2.5 \\ 0.2x + 0.5 & \text{if } -2.5 < x < 2.5 \\ 1 & \text{if } x \geq 2.5 \end{cases}$$"
+        elif func_name == "swish":
+            sigmoid = 1 / (1 + np.exp(-x))
+            y = x * sigmoid
+            formula = r"$$Swish(x) = x \times \sigma(x)$$"
+        elif func_name == "mish":
+            y = x * np.tanh(np.log1p(np.exp(x)))
+            formula = r"$$Mish(x) = x \times tanh(ln(1 + e^x))$$"
+        
+        plt.figure(figsize=(10, 6))
+        plt.plot(x, y)
+        plt.grid(True)
+        plt.axhline(y=0, color='k', linestyle='-', alpha=0.3)
+        plt.axvline(x=0, color='k', linestyle='-', alpha=0.3)
+        plt.title(f'{func_name} Activation Function')
+        plt.xlabel('x')
+        plt.ylabel('y')
+        return plt.gca(), formula
+
+    mo.md(
+        f"""
+        {plot_activation(activation_functions.value)[1]}
+        {mo.as_html(plot_activation(activation_functions.value)[0])}
+        """
+    )
+    return np, plot_activation, plt
+
+
+@app.cell(hide_code=True)
+def __(mo):
+    activation_functions = mo.ui.dropdown(
```

Comment:
I actually think this can be a multiselect.
Then readers can select multiple activation functions at the same time and visualize their differences clearly

Reply:
That's a nice suggestion; also, can we also try experimenting with the columns feature (wasim) introduced? I guess multi-select/column approach would be a good addition.

Reply:
That's still an experimental feature right? Not sure whether it will work well on stable marimo versions

Reply:
> That's still an experimental feature right? Not sure whether it will work well on stable marimo versions

Ahh right; I always wanted to use that feature ever since it came out; guess I can try it locally to see how it fares.